src/nbl/ext/DebugDraw/CDrawAABB.cpp

```diff
@@ -0,0 +1,374 @@
+		std::make_pair(binding.memory->isCurrentlyMapped(), "Streaming buffer's memory must be mapped!"), // streaming buffer contructor already validates it, but cannot assume user won't unmap its own buffer for some reason (sorry if you have just hit it)
+		std::make_pair(binding.memory->getCurrentMappingAccess().hasFlags(getRequiredAccessFlags(binding.memory->getMemoryPropertyFlags())), "Streaming buffer's memory current mapping access flags don't meet requirements!")
+		});
+
+	for (const auto& [ok, error] : validation)
+		if (!ok)
+		{
+			params.utilities->getLogger()->log(error, ILogger::ELL_ERROR);
+			return false;
+		}
+
+	return true;
+}
+
+core::smart_refctd_ptr<video::IGPUPipelineLayout> DrawAABB::createDefaultPipelineLayout(video::ILogicalDevice* device, const asset::SPushConstantRange& pcRange)
+{
+	return device->createPipelineLayout({ &pcRange , 1 }, nullptr, nullptr, nullptr, nullptr);
+}
+
+core::smart_refctd_ptr<video::IGPUPipelineLayout> DrawAABB::createDefaultPipelineLayout(video::ILogicalDevice* device)
+{
+	SPushConstantRange pcRange = {
+		.stageFlags = IShader::E_SHADER_STAGE::ESS_VERTEX,
+		.offset = 0,
+		.size = sizeof(SPushConstants)
+	};
+	return device->createPipelineLayout({ &pcRange , 1 }, nullptr, nullptr, nullptr, nullptr);
+}
+
+smart_refctd_ptr<IGPUGraphicsPipeline> DrawAABB::createDefaultPipeline(video::ILogicalDevice* device, video::IGPUPipelineLayout* layout, video::IGPURenderpass* renderpass, video::IGPUGraphicsPipeline::SShaderSpecInfo& vertex, video::IGPUGraphicsPipeline::SShaderSpecInfo& fragment)
+{
+	smart_refctd_ptr<video::IGPUGraphicsPipeline> pipeline;
+
+	video::IGPUGraphicsPipeline::SCreationParams params[1] = {};
+	params[0].layout = layout;
+	params[0].vertexShader = vertex;
+	params[0].fragmentShader = fragment;
+	params[0].cached = {
+		.primitiveAssembly = {
+			.primitiveType = asset::E_PRIMITIVE_TOPOLOGY::EPT_LINE_LIST,
+		}
+	};
+	params[0].renderpass = renderpass;
+
+	device->createGraphicsPipelines(nullptr, params, &pipeline);
+
+	return pipeline;
+}
+
+bool DrawAABB::renderSingle(IGPUCommandBuffer* commandBuffer)
+{
+	commandBuffer->setLineWidth(1.f);
+	commandBuffer->draw(24, 1, 0, 0);
+
+	return true;
+}
+
+bool DrawAABB::render(IGPUCommandBuffer* commandBuffer, ISemaphore::SWaitInfo waitInfo, float* cameraMat3x4)
+{
+	using offset_t = SCachedCreationParameters::streaming_buffer_t::size_type;
+	constexpr auto MdiSizes = std::to_array<offset_t>({ sizeof(float32_t3), sizeof(InstanceData) });
+	// shared nPoT alignment needs to be divisible by all smaller ones to satisfy an allocation from all
+	constexpr offset_t MaxAlignment = std::reduce(MdiSizes.begin(), MdiSizes.end(), 1, [](const offset_t a, const offset_t b)->offset_t {return std::lcm(a, b); });
+	// allocator initialization needs us to round up to PoT
+	const auto MaxPOTAlignment = roundUpToPoT(MaxAlignment);
```

Comment:
you shall only allocate the instance data in the MDI buffer, it will make all the logic here a lot simpler!

Index buffer must be premade and constant at `create` time